### DISCUSSION TITLE
Filament Advanced Forms

### DISCUSSION THREAD

**Discussion by MasterAdmin2** [filamentphp/filament#11931] [Help] [answered]
### Package

Form builder

### Package Version

3.2

### How can we help you?

Sir,
I just need simple support from your side if possible, i am using filament and trying to calculate, but it is not working for me, i know I am making a silly mistake but unable to catch that, can you please help me, if possible, I know it won't take long
Also, adding the source code for my form,

> 
`` public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Section::make('Input Variables')
                    ->description('Prevent abuse by limiting the number of requests per period')
                    ->columns(3)
                    ->schema([
                        Select::make('customer_id')
                            ->relationship('customer', 'name')
                            ->required(),
                        DatePicker::make('date')
                            ->minDate('01-01-1990')
                            ->maxDate(today())
                            ->required(),
                        TextInput::make('vehicle_no')
                            ->maxLength(10),
                    ]),
                Section::make('Input Variables')
                    ->description('Prevent abuse by limiting the number of requests per period')
                    ->columns(4)
                    ->schema([
                        TextInput::make('milk_weight')
                            ->required()
                            ->numeric()
                            ->live(onBlur: true),
                        TextInput::make('rate')
                            ->required()
                            ->numeric()
                            ->live(onBlur: true)
                            ->afterStateUpdated(function ($set, $state) {
                                $rate = (float)$state;
                                $set('fat_rate', number_format((($rate * 100) / 6.5) * 0.6, 2));
                                $set('snf_rate', number_format((($rate * 100) / 8.5) * 0.4, 2));
                            }),
                        TextInput::make('fat')
                            ->required()
                            ->numeric()
                            ->live(onBlur: true)
                            ->afterStateUpdated(function ($set, $state) {
                                $milk_weight = (float)$state;
                                $fat = (float)$state;
                                $fat_kg = number_format(($fat * $milk_weight) / 1000, 2);
                                $set('fat_kg', $fat_kg);
                            }),
                        TextInput::make('clr')
                            ->required()
                            ->numeric()
                            ->live(onBlur: true)
                            ->afterStateUpdated(function ($set, $state) {
                                $clr = (float)$state;
                                $fat = (float)$state;
                                $set('snf_percentage', number_format((($clr/4) + 0.14 + ($fat*2)) / 100, 2));

                                $fat_kg = (float)$state;
                                $fat_rate = (float)$state;
                                $set('fat_amount', number_format(($fat_kg * $fat_rate), 2));
                            }),
                    ]),
                Section::make('Auto Calculated Fields')
                    ->columns(4)
                    ->schema([
                        TextInput::make('fat_rate')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500),
                        TextInput::make('fat_kg')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500),
                        TextInput::make('fat_amount')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500),
                        TextInput::make('snf_kg')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500),
                        TextInput::make('snf_rate')
                           ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500),
                        TextInput::make('snf_percentage')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500),
                        TextInput::make('snf_amount')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500),
                        TextInput::make('grand_total')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500),
                    ]),
            ]);
    }``
> 

they are giving me the wrong output on automatic calculation,

the only right calculation is from 2 fields based on rate,

can anyone @leandrocfe  help me solve this problem,

**Comment by leandrocfe** [comment]
What is the issue?

**Reply by MasterAdmin2** [reply]
i solved most of it by using separate variables, but it is calculating wrong, as I checked on excel, it is giving me wight calc, but when I am using it inside it, it is showing different result.
my piece of code, 
if u want I can share excel as well.



`$rateUpdate = function (Set $set, Get $get, $state) {
            $set('rate', (float)$state); // Store rate in the state
        };
        $milkWeightUpdate = function (Set $set, Get $get, $state) {
            $set('milk_weight', (float)$state); // Store milk_weight in the state
        };
        $fatUpdate = function (Set $set, Get $get, $state) {
            $set('fat', (float)$state); // Store fat in the state
        };
        $clrUpdate = function (Set $set, Get $get, $state) {
            $set('clr', (float)$state); // Store clr in the state
        };
        $fatRateUpdate = function (Set $set, Get $get, $state) {
            $set('fat_rate', (float)$state); // Store fatRate in the state
        };
        $fatKgUpdate = function (Set $set, Get $get, $state) {
            $set('fat_kg', (float)$state); // Store fatKg in the state
        };
        $snfPercentageUpdate = function (Set $set, Get $get, $state) {
            $set('snf_percentage', (float)$state); // Store snfPercentage in the state
        };
        $snfKgUpdate = function (Set $set, Get $get, $state) {
            $set('snf_kg', (float)$state); // Store snfPercentage in the state
        };
        $fatAmountUpdate = function (Set $set, Get $get, $state) {
            $set('fat_amount', (float)$state); // Store snfPercentage in the state
        };
        $snfAmountUpdate = function (Set $set, Get $get, $state) {
            $set('snf_amount', (float)$state); // Store snfPercentage in the state
        };

        // Define a callback function to calculate the result based on updated
        $fatRateCalculator = function (Set $set, Get $get) {
            $rate = (float)$get('rate');
            $fat_rate = (number_format((($rate * 100) / 6.5) * 0.6, 2));
            $set('fat_rate', $fat_rate); // Update the result field with the calculated value
        };
        $snfRateCalculator = function (Set $set, Get $get) {
            $rate = (float)$get('rate');
            $snf_rate = number_format((($rate * 100) / 8.5) * 0.4, 2);
            $set('snf_rate', $snf_rate); // Update the result field with the calculated value
        };
        $fatKgCalculator = function (Set $set, Get $get) {
            $milk_weight = (float)$get('milk_weight');
            $fat = (float)$get('fat');
            $fat_kg = ($milk_weight * $fat) / 1000;
            $set('fat_kg', $fat_kg); // Update the result field with the calculated value
        };
        $fatAmountCalculator = function (Set $set, Get $get) {
            $fat_rate = (float)$get('fat_rate');
            $fat_kg = (float)$get('fat_kg');
            $fat_amount = $fat_rate * $fat_kg;
            $set('fat_amount', $fat_amount); // Update the result field with the calculated value
        };
        $snfPercentageCalculator = function (Set $set, Get $get) {
            $fat = (float)$get('fat');
            $clr = (float)$get('clr');
            $snf_percentage = number_format(($clr / 4) + 0.14 + ($fat * 2) / 100, 2);
            $set('snf_percentage', $snf_percentage); // Update the result field with the calculated value
        };
        $snfKgCalculator = function (Set $set, Get $get) {
            $milk_weight = (float)$get('milk_weight');
            $snf_percentage = (float)$get('snf_percentage');
            $snf_kg = ($milk_weight * $snf_percentage) / 100;
            $set('snf_kg', $snf_kg); // Update the result field with the calculated value
        };
        $snfAmountCalculator = function (Set $set, Get $get) {
            $snf_kg = (float)$get('snf_kg');
            $snf_rate = (float)$get('snf_rate');
            $snf_amount = ($snf_kg * $snf_rate);
            $set('snf_amount', $snf_amount); // Update the result field with the calculated value
        };
        $grandTotalCalculator = function (Set $set, Get $get) {
            $fat_amount = (float)$get('fat_amount');
            $snf_amount = (float)$get('snf_amount');
            $grand_total = ($fat_amount + $snf_amount);
            $set('grand_total', $grand_total); // Update the result field with the calculated value
        };`


inside it, fat rate and snf rate is miscalculated, everything else is correct.

**Reply by MasterAdmin2** [reply]
and this is the form schema 

`
->schema([
                Section::make('Input Variables')
                    ->description('Prevent abuse by limiting the number of requests per period')
                    ->columns(3)
                    ->schema([
                        Select::make('customer_id')
                            ->relationship('customer', 'name')
                            ->required(),
                        DatePicker::make('date')
                            ->minDate('01-01-1990')
                            ->maxDate(today())
                            ->required(),
                        TextInput::make('vehicle_no')
                            ->maxLength(10),
                    ]),
                Section::make('Input Variables')
                    ->description('Prevent abuse by limiting the number of requests per period')
                    ->columns(4)
                    ->live(onBlur: true)
                    ->debounce(1500)
                    ->schema([
                        TextInput::make('rate')
                            ->required()
                            ->numeric()
                            ->live(onBlur: true)
                            ->afterStateUpdated($rateUpdate),
                        TextInput::make('milk_weight')
                            ->required()
                            ->numeric()
                            ->live(onBlur: true)
                            ->afterStateUpdated($milkWeightUpdate),
                        TextInput::make('fat')
                            ->required()
                            ->numeric()
                            ->live(onBlur: true)
                            ->afterStateUpdated($fatUpdate),
                        TextInput::make('clr')
                            ->required()
                            ->numeric()
                            ->live(onBlur: true)
                            ->afterStateUpdated($clrUpdate)
                            ->afterStateUpdated($fatRateCalculator)
                            ->afterStateUpdated($snfRateCalculator)
                            ->afterStateUpdated($fatKgCalculator)
                            ->afterStateUpdated($fatAmountCalculator)
                            ->afterStateUpdated($snfPercentageCalculator)
                            ->afterStateUpdated($snfKgCalculator)
                            ->afterStateUpdated($snfAmountCalculator)
                            ->afterStateUpdated($grandTotalCalculator),
                    ]),
                Section::make('Auto Calculated Fields')
                    ->columns(4)
                    ->schema([
                        TextInput::make('fat_rate')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500)
                            ->afterStateUpdated($fatRateCalculator),
                        TextInput::make('snf_rate')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500)
                            ->afterStateUpdated($snfRateCalculator)
                            ->afterStateUpdated($fatRateUpdate),
                        TextInput::make('fat_kg')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500)
                            ->afterStateUpdated($fatKgCalculator)
                            ->afterStateUpdated($fatKgUpdate),
                        TextInput::make('fat_amount')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500)
                            ->afterStateUpdated($fatAmountCalculator)
                            ->afterStateUpdated($fatAmountUpdate),
                        TextInput::make('snf_percentage')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500)
                            ->afterStateUpdated($snfPercentageCalculator)
                            ->afterStateUpdated($snfPercentageUpdate),
                        TextInput::make('snf_kg')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->afterStateUpdated($snfKgCalculator)
                            ->afterStateUpdated($snfKgUpdate)
                            ->debounce(1500),
                        TextInput::make('snf_amount')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->debounce(1500)
                            ->afterStateUpdated($snfAmountUpdate),
                        TextInput::make('grand_total')
                            ->numeric()
                            ->live(onBlur: true)
                            ->readOnly()
                            ->afterStateUpdated($grandTotalCalculator)
                            ->debounce(1500),
                    ]),
            ]);
`

**Reply by leandrocfe** [reply, marked as answer]
_but it is calculating wrong..._
What is wrong? What is the issue?
